src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -2377,15 +2370,21 @@ private unsafe void CreateUriInfo(Flags cF)
 
         Done:
             cF |= Flags.MinimalUriInfoSet;
-            info.DnsSafeHost = _dnsSafeHost;
+
+            Debug.Assert(sizeof(Flags) == sizeof(ulong));
 
             Interlocked.CompareExchange(ref _info, info, null!);
-            lock (_info)
+
+            Flags current = _flags;
+            while ((current & Flags.MinimalUriInfoSet) == 0)
             {
-                if ((_flags & Flags.MinimalUriInfoSet) == 0)
+                Flags newValue = (current & ~Flags.IndexMask) | cF;
+                ulong oldValue = Interlocked.CompareExchange(ref Unsafe.As<Flags, ulong>(ref _flags), (ulong)newValue, (ulong)current);
+                if (oldValue == (ulong)current)
                 {
-                    _flags = (_flags & ~Flags.IndexMask) | cF;
+                    return;
                 }
+                current = (Flags)oldValue;
```

Comment:
This could alternatively be an if/else block based on whether CompareExchange returned null
**if**: a regualar update of the flags field
**else**: spin lock waiting for MinimalUriInfoSet to be set